projects/packages/waf/tests/php/integration/test-waf-activation.php

```diff
@@ -80,4 +104,79 @@ public function testActivation() {
 		remove_filter( 'pre_http_request', array( $this, 'return_sample_response' ) );
 	}
 
+	/**
+	 * Test WAF deactivation.
+	 */
+	public function testDeactivation() {
+		$deactivated = Waf_Initializer::on_deactivation();
+
+		// Ensure the WAF was deactivated successfully.
+		$this->assertTrue( $deactivated );
+
+		// Ensure the options were deleted.
+		$this->assertSame( get_option( Waf_Runner::SHARE_DATA_OPTION_NAME ), false );
+		$this->assertSame( get_option( Waf_Runner::MODE_OPTION_NAME ), false );
+
+		// Ensure the rules entrypoint file was emptied.
+		$this->assertSame( file_get_contents( Waf_Runner::get_waf_file_path( Waf_Rules_Manager::RULES_ENTRYPOINT_FILE ) ), "<?php\n" );
+	}
+
+	/**
+	 * Test WAF activation when the filesystem is unavailable.
+	 */
+	public function testActivationReturnsWpErrorWhenFilesystemUnavailable() {
+		// Mock the WPCOM request for retrieving the automatic rules.
+		add_filter( 'pre_http_request', array( $this, 'return_sample_response' ) );
+
+		// Break the filesystem.
+		add_filter( 'filesystem_method', array( $this, 'return_invalid_filesystem_method' ) );
+
+		// Initialize the firewall.
+		$activated = Waf_Initializer::on_activation();
+
+		// Validate the error.
+		$this->assertTrue( is_wp_error( $activated ) );
+		$this->assertSame( 'file_system_error', $activated->get_error_code() );
+
+		// Clean up.
+		remove_filter( 'pre_http_request', array( $this, 'return_sample_response' ) );
+		remove_filter( 'filesystem_method', array( $this, 'return_invalid_filesystem_method' ) );
```

Comment:
These might be better to go into the `tearDown()` method because the way PHPunit works is that it throws an exception (internally) of some assertion fails, like the `assertSame()` above, which would mean that these calls will never be reached. This would then in turn possibly affect the outcome of other following tests to be executed.

Having said that, as long as one test fails, the tests should be (and are) certainly considered a failure in total. It might becloud the failure report though.

Disclaimer: I have not compared to how other projects in the monorepo do it, so if this is the default approach already, totally fine!

Reply:
I explored the monorepo looking at uses of `remove_filter()` in `test-*.php` files and it looks like filter clean up inside the tests is the most common approach.

One interesting implementation I found was:

```php
add_filter( /* ... */ );
try {
    $this->assertTrue( /* ... */ );
} finally {
    remove_filter( /* ... */ );
}
```

Reply:
Ah, turns out WorDBless cleans up for us!

```php
/**
 * After a test method runs, reset any state in WordPress the test method might have changed.
 *
 * @after
 */
public function tear_down_wordbless() {
    $this->_restore_hooks();
    Options::init()->clear_options();
    Posts::init()->clear_all_posts();
    PostMeta::init()->clear_all_meta();
    Users::init()->clear_all_users();
    $this->clear_uploads();
}
```

Reply:
Very nice! :) Thanks for checking!